### PR TITLE
Next batch of ngtsc changes

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/util.ts [hunk 1]
```diff
@@ -87,3 +87,10 @@ export function referenceToExpression(ref: Reference, context: ts.SourceFile): E
 export function isAngularCore(decorator: Decorator): boolean {
   return decorator.import !== null && decorator.import.from === '@angular/core';
 }
+
+export function unwrapExpression(node: ts.Expression): ts.Expression {
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/util.ts]: add a one liner comment ?

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 2]
```diff
@@ -125,6 +127,46 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
       },
     ];
   }
+
+  /**
+   * Given a `FunctionDeclaration` or `MethodDeclaration`, check if it is typed as a
+   * `ModuleWithProviders` and return an expression referencing the module if available.
+   */
+  private _resolveModuleWithProviders(node: ts.FunctionDeclaration|
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: IMO the name doesn't reflect what the method does

File: packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts [hunk 3]
```diff
@@ -179,9 +181,14 @@ export class AbsoluteReference extends Reference {
  * @param checker a `ts.TypeChecker` used to understand the expression
  * @returns a `ResolvedValue` representing the resolved value
  */
-export function staticallyResolve(node: ts.Expression, checker: ts.TypeChecker): ResolvedValue {
-  return new StaticInterpreter(checker).visit(
-      node, {absoluteModuleName: null, scope: new Map<ts.ParameterDeclaration, ResolvedValue>()});
+export function staticallyResolve(
+    node: ts.Expression, checker: ts.TypeChecker,
+    foreignFunctionResolver?: (node: ts.FunctionDeclaration | ts.MethodDeclaration) =>
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: pls update the api doc

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 4]
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
...
+  const first = name === 'ViewChild' || name === 'ContentChild';
+  const arg = staticallyResolve(args[0], checker);
+
+  // Extract the predicate
+  let predicate: Expression|string[]|null = null;
+  if (arg instanceof Reference) {
+    predicate = new WrappedNodeExpr(args[0]);
+  } else if (typeof arg === 'string') {
+    predicate = [arg];
+  } else if (assertIsStringArray(arg, '@' + name)) {
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: `assert` are often debug time checks only, rename/refactor ?

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 5]
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
...
+    predicate = arg as string[];
+  }
+  if (predicate === null) {
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: should it be a final `else {}` ?

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 6]
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
...
+    if (!ts.isObjectLiteralExpression(optionsExpr)) {
+      throw new Error(`@${name} options must be an object literal`);
+    }
+    const options = reflectObjectLiteral(optionsExpr);
+    if (options.has('read')) {
+      read = new WrappedNodeExpr(options.get('read') !);
+    }
+
+    if (options.has('descendants')) {
+      const descendantsValue = staticallyResolve(options.get('descendants') !, checker);
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: nit: remove the local variable and use `descendants` ?

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 7]
```diff
@@ -112,29 +132,112 @@ export function extractDirectiveMetadata(
...
+  } else if (assertIsStringArray(arg, '@' + name)) {
+    predicate = arg as string[];
+  }
+  if (predicate === null) {
+    throw new Error(`@${name} predicate cannot be interpreted`);
+  }
+
+  // Extract the read and descendants options.
+  let read: Expression|null = null;
+  let descendants: boolean = true;
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: shouldn't `descendants` be a function of `first` ? (I think the behavior is different, can't remember exactly)

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 8]
```diff
@@ -199,3 +302,109 @@ function parseDecoratedFields(
...
+            if (typeof resolved !== 'string') {
+              throw new Error(`@HostListener()'s event name argument must be a string`);
+            }
+
+            eventName = resolved;
+
+            if (decorator.args.length === 2) {
+              const resolvedArgs = staticallyResolve(decorator.args[1], checker);
+              if (!assertIsStringArray(resolvedArgs, '@HostListener.args')) {
+                throw new Error(`@HostListener arguments must be a string array`);
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: nit: `arguments` -> `second argument` would be clearer IMO

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 9]
```diff
@@ -412,4 +412,64 @@ describe('ngtsc behavioral tests', () => {
...
+          selector: 'test',
+          template: '<div #foo></div>',
+          queries: {
+            'mview': new ViewChild('test1'),
+            'mcontent': new ContentChild('test2'),
+          }
+        })
+        class FooCmp {
+          @ContentChild('bar', {read: TemplateRef}) child: any;
+          @ContentChildren(TemplateRef) children: any;
```
Comment (vicb) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Add tests for `@View...` and getters / setters ?

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 10]
```diff
@@ -412,4 +412,64 @@ describe('ngtsc behavioral tests', () => {
...
+    writeConfig();
+    write(`test.ts`, `
+        import {Component, ContentChild, ContentChildren, TemplateRef} from '@angular/core';
+
+        @Component({
+          selector: 'test',
+          template: 'Test',
+          host: {
+            '[attr.hello]': 'foo',
+            '(click)': 'onClick($event)',
```
Comment (vicb) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: add test for args ?

File: packages/compiler-cli/src/ngtsc/transform/src/translator.ts [hunk 11]
```diff
@@ -6,10 +6,19 @@
...
+import {ArrayType, AssertNotNull, BinaryOperator, BinaryOperatorExpr, BuiltinType, BuiltinTypeName, CastExpr, ClassStmt, CommaExpr, CommentStmt, ConditionalExpr, DeclareFunctionStmt, DeclareVarStmt, Expression, ExpressionStatement, ExpressionType, ExpressionVisitor, ExternalExpr, ExternalReference, FunctionExpr, IfStmt, InstantiateExpr, InvokeFunctionExpr, InvokeMethodExpr, JSDocCommentStmt, LiteralArrayExpr, LiteralExpr, LiteralMapExpr, MapType, NotExpr, ReadKeyExpr, ReadPropExpr, ReadVarExpr, ReturnStatement, Statement, StatementVisitor, StmtModifier, ThrowStmt, TryCatchStmt, Type, TypeVisitor, WrappedNodeExpr, WriteKeyExpr, WritePropExpr, WriteVarExpr} from '@angular/compiler';
 import * as ts from 'typescript';
+
 import {relativePathBetween} from '../../util/src/path';
 
+export class Context {
+  constructor(readonly isStatement: boolean) {}
+
+  get setExpressionMode(): Context { return this.isStatement ? new Context(false) : this; }
```
Comment (vicb) [inline packages/compiler-cli/src/ngtsc/transform/src/translator.ts]: nit: weird that a getter as side effect. make it a regular method ?

File: packages/compiler-cli/src/ngtsc/annotations/src/util.ts (hunk 1 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/util.ts]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Renamed to `_extractModuleFromModuleWithProvidersFn`

**Review by vicb — APPROVED** [review]
Only a few minor comments, mostly need.

Seeing the size of this PR frightened me at first but HUGE kudos to you for:
- Splitting the PR in small and scoped commits,
- Adding exhaustive PR comments to help understanding where it makes sense,
- Adding inline docs were it makes sense.

It turned out to be a very pleasant PR to review - thanks for easing the review.

File: packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: Done.

File: packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts [hunk 12]
```diff
@@ -356,6 +358,22 @@ class StaticInterpreter {
     return this.visitSymbol(symbol, context);
   }
 
+  private visitTemplateExpression(node: ts.TemplateExpression, context: Context): ResolvedValue {
+    const pieces: string[] = [node.head.text];
+    for (let i = 0; i < node.templateSpans.length; i++) {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: Tagged templates are not handled yet, so they'll fall through and return `DYNAMIC_VALUE`. I don't even know if `ngc` handles them.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Renamed to `isStringArrayOrDie()`

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 6 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: The assignment isn't legal until after the `typeof` check.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 13]
```diff
@@ -80,17 +80,19 @@ export function extractDirectiveMetadata(
   const decoratedElements =
       members.filter(member => !member.isStatic && member.decorators !== null);
 
+  const coreModule = isCore && '@angular/core' || undefined;
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 14]
```diff
@@ -109,15 +109,29 @@ export class ComponentDecoratorHandler implements DecoratorHandler<R3ComponentMe
...
     // If the component has a selector, it should be registered with the `SelectorScopeRegistry` so
     // when this component appears in an `@NgModule` scope, its selector can be determined.
-    if (directiveMetadata.selector !== null) {
-      this.scopeRegistry.registerSelector(node, directiveMetadata.selector);
+    if (metadata.selector !== null) {
+      this.scopeRegistry.registerSelector(node, metadata.selector);
+    }
+
+    // Construct the list of view queries.
+    const coreModule = this.isCore && '@angular/core' || undefined;
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 5 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 7 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 8 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 15]
```diff
@@ -109,15 +112,31 @@ export class ComponentDecoratorHandler implements DecoratorHandler<R3ComponentMe
...
     // If the component has a selector, it should be registered with the `SelectorScopeRegistry` so
     // when this component appears in an `@NgModule` scope, its selector can be determined.
-    if (directiveMetadata.selector !== null) {
-      this.scopeRegistry.registerSelector(node, directiveMetadata.selector);
+    if (metadata.selector !== null) {
+      this.scopeRegistry.registerSelector(node, metadata.selector);
+    }
+
+    // Construct the list of view queries.
+    const coreModule = this.isCore ? '@angular/core' : undefined;
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: Hmm, during core compilation, `coreModule` should be undefined right? This seems inverted now.

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**